### ISSUE TITLE
Getting the pickle analysis data from within Python

### ISSUE THREAD

**Issue by JohnnyRacer** [huggingface/huggingface_hub#1212]
Hello, I am trying to obtain the pickle analysis data that is associated with pickled model checkpoints from Python programmatically but is unable to find anything in the docs on how to do this. Is there a clear way to do this?

A screenshot of the example of the analysis information is attached below. 
![pickle](https://user-images.githubusercontent.com/77214388/202942853-75730347-55e0-4fdc-a651-eb478c303920.png)

**Comment by julien-c** [comment]
it is not currently supported, but cc'ing @McPatate (and @krampstudio for visibility)

thanks for the feature request!

**Comment by JohnnyRacer** [comment]
@Wauplin  Thanks a lot for adding it as a new feature, just tested out the new API and it works very well! I was wondering if in the future a user downloading a potential malicious Pickle file would raise a confirmation warning to make it clear to them the risks before allowing the model to load? This seems like a good addition now that the scan data is available via the API.

**Comment by Wauplin** [comment]
>  I was wondering if in the future a user downloading a potential malicious Pickle file would raise a confirmation warning to make it clear to them the risks before allowing the model to load? 

That's actually a pretty good idea. I'll keep it in mind to do it properly. I think that if we do something like that, it will be a warning but without asking confirmation from the user (or at least not by default). Always a balance to find between usability and security :)

**Comment by McPatate** [comment]
I think having an extra option `disable_unsafe_download` (defaulting to `False`) could be a nice to have 😄

**Comment by Wauplin** [comment]
Yes, I was thinking about an environment variable directly so that users can be sure to disable them globally.

**Comment by Wauplin** [comment]
@McPatate before implementing this, what's your take on how to consider a file as "unsafe". I saw that the security scan has 3 levels: `innocuous`, `suspicious` and `dangerous`.

I see four options here:
1. set `HF_HUB_DISABLE_UNSAFE_DOWNLOAD=1` raise an Exception on dangerous files but not suspicious ones
2. set `HF_HUB_DISABLE_UNSAFE_DOWNLOAD=1` raise an Exception on dangerous files AND suspicious ones
3. set `HF_HUB_DISABLE_DANGEROUS_DOWNLOAD=1` and `HF_HUB_DISABLE_SUSPICIOUS_DOWNLOAD=1` separately. If "disable suspicious" is True, "disable dangerous" is automatically True as well => more explicit and flexible but it adds two env variables instead of one.
4. set `HF_HUB_UNSAFE_DOWNLOAD_LEVEL` (or `HF_HUB_SAFETY_LEVEL`?) to either "minimal" (default), "intermediate" or "maximal" => as much flexibility as 3. but with only 1 variable

For me it really depends on what `suspicious` means and when it happens. IMO if a user sets `HF_HUB_DISABLE_UNSAFE_DOWNLOAD=1`,  I'd rather forbid any non-innocuous download (i.e. option 2). Otherwise if flexibility is important, I'd go for option 4 but with a better naming.

Also cc @JohnnyRacer if you are opinionated on this topic.

**Comment by JohnnyRacer** [comment]
@Wauplin 

I think it would be a good idea to prevent downloads of  dangerous Pickle checkpoints by default, with  a warning message that can be disabled via an environment variable like you had mentioned above or by calling a function like `transformers.env_configs.disable_unsafe_download(True) .  To toggle the ability to allow for unsafe download on or off. I think a way to disable unsafe downloads is needed as it can be used conveniently as a malware dropper, especially when models are loaded onto production environments.  For suspicious models I think a toggle-able  warning will suffice since most do not contain dangerous `eval` which would allow for arbitrary code execution.

**Comment by julien-c** [comment]
I think the future-proof way is to emit a warning on _any_ model loading from Pickle, as we default more and more to the safetensors format (and its usage is growing rapidly)

**Comment by McPatate** [comment]
Does it cost a lot to send a request to the hub to check before starting a download if there are dangerous imports?

It could also be done asynchronously, as in no changes in the download logic, except you add a cancellation token to the download job and if the file contains a dangerous import, the asynchronous job cancels the download, wdyt @julien-c?

**Comment by julien-c** [comment]
> But that should rather live in libraries

Yes, correct

**Comment by julien-c** [comment]
or even in torch

**Comment by jbbqqf** [comment]
Hi! Triaging older issues — the original ask here (programmatic access to the pickle scan data) has shipped.

On current `main`, scan results are exposed on each `RepoFile`:

- [`src/huggingface_hub/hf_api.py#L454-L462`](https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/hf_api.py#L454-L462) — `BlobSecurityInfo` dataclass with `pickle_import_scan` and `av_scan` fields.
- [`src/huggingface_hub/hf_api.py#L788-L797`](https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/hf_api.py#L788-L797) — the field is populated from the Hub's `securityFileStatus.pickleImportScan` payload, so you get it on any `RepoFile` returned by `HfApi.list_repo_tree`, `paths_info`, `model_info(files_metadata=True)`, etc.

So `repo_file.security.pickle_import_scan` is your hook.

The later discussion in this thread (download-time blocking on unsafe pickles via `HF_HUB_DISABLE_UNSAFE_DOWNLOAD`, etc.) is a separate concern that the team noted would belong upstream in `transformers`/`diffusers` rather than in `huggingface_hub` itself. If you'd still like that piece tracked, I'd suggest a fresh issue scoped to just the env-var/download policy — but the data-access ask in the original report is solved.

Would you mind closing this out?

(Disclosure: I drafted this comment with help from Claude Code while triaging stale issues; the references above were verified manually against current `main` at 2003e8b.)

**Comment by Wauplin** [comment]
I'll close rather this one as "won't fix"
